public/app/plugins/datasource/influxdb/datasource.ts

```diff
@@ -166,11 +168,41 @@ export default class InfluxDatasource extends DataSourceWithBackend<InfluxQuery,
     }
   }
 
+  async getRetentionPolicies(): Promise<string[]> {
+    if (this.retentionPolicies.length) {
+      return Promise.resolve(this.retentionPolicies);
+    } else {
+      return getAllPolicies(this);
+    }
+  }
+
   query(request: DataQueryRequest<InfluxQuery>): Observable<DataQueryResponse> {
     if (!this.isProxyAccess) {
       const error = new Error(BROWSER_MODE_DISABLED_MESSAGE);
       return throwError(() => error);
     }
+
+    // When the dashboard first load or on dashboard panel edit mode
+    // PanelQueryRunner runs the queries to have a visualization on the panel.
+    // At that point datasource doesn't have the retention policies fetched.
+    // So hardcoded policy is being sent. Which causes problems.
+    // To overcome this we check/load policies first and then do the query.
+    return defer(() => this.getRetentionPolicies()).pipe(
```

Comment:
My only concern is that this does not handle errors, if anything were to throw in the getRetentionPolicies (and there are promises resolving in there!) this would break.
<img width="1728" alt="image" src="https://user-images.githubusercontent.com/109082771/231824837-df0df3aa-ea79-439d-85b6-2a03e48cd08e.png">

Can we make sure that `getRetentionPolicies` cannot emit uncaught exceptions?
```
 async getRetentionPolicies(): Promise<string[]> {
    if (this.retentionPolicies.length) {
      return Promise.resolve(this.retentionPolicies);
    } else {
      return getAllPolicies(this).catch(err => {
        console.error('Unable to fetch retention policies, stuff might break', err); //@todo better error message?
        return Promise.resolve(this.retentionPolicies)
      })
    }
  }
```

Reply:
Makes sense. I will make the change. Thanks for the suggestion.

Reply:
Do you have a better error message suggestion?
I would go `Unable to fetch retention policies` for now.